# misFinder

Identify mis-assemblies in an unbiased manner using reference and paired-end reads

============================

The main features of misFinder can be summarized as follows:

* misFinder is a tool that aims to identify the assembly errors with high accuracy in an unbiased way and correct these errors at their mis-assembled positions to improve the assembly accuracy for downstream analysis. It combines the information of reference (or close related reference) genome and aligned paired-end reads to the assembled sequence. Structure variation and mis-assembly can be detected by comparing the reference genome and assembled sequence. Different types of assembly errors can then be distinguished from the mis-assembled sequences by analyzing the aligned paired-end reads using multiple features derived from coverage and consistence of insert distance to obtain high confident error calls.
* misFinder is based on BLASTN (version 2.2.25+ and higher), and assembly (contigs/scaffolds), genome reference and paired-end reads from Solexa sequencing technology are its input. BLASTN can be downloaded from http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download, and then the BLASTN command 'blastn' should be added into to $PATH or into one of the following directories: /usr/local/bin, /usr/bin, /bin, /usr/local/sbin, /usr/sbin, /sbin, $HOME/bin.
* There is a configuration file to configure the assembly, reference and paired-end reads information before running misFinder. The configuration file can be found in the downloaded folder.
* misFinder can accepte the paired-end reads in fastq or fasta file format, and these paired-end reads can be stored interleaved in a single file (paired_reads.fastq) or stored in two separate files (reads_1.fastq and reads_2.fastq). Reads can be all the same length or different lengths.


## Prerequisites

misFinder depends on the following libraries and tools:
* BLASTN (http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download)
* g++ (v4.7 or higher which supports c++11)


## Compiling misFinder

The binary file `mf` can be generated by typing:
```sh
$ wget -c https://github.com/zhuxiao/misFinder/releases/download/0.5.0/misfinder_0.5.0.tar.xz
$ tar -xf misfinder_0.5.0.tar.xz
$ cd misfinder_0.5.0/
./autogen.sh
```
and the binary file 'mf' will be output into the folder 'bin'.

Besides, it is recommanded to create the link of `mf` in `bin` directory to the `$PATH` directory.
```sh
# create soft-link of mf to '~/bin' which should be already exist or create a new one
$ mkdir ~/bin
$ ln -s $PWD/bin/mf ~/bin
# or, create soft-link of mf to /usr/local/bin, which needs the sudo permission
$ sudo ln -s $PWD/bin/mf /usr/local/bin
```


### Compiling Prerequisites

Compling prerequisites:

* Ubuntu:
```sh
$ sudo apt install g++
```

* CentOS:
```sh
$ sudo dnf install gcc-c++
```


## Quick Start

If the assembly, reference and paired-end reads information are configured in the configuration file, the command will be:
```sh
# detect mis-assemblies based on the specified information in 'config' file
$ mf all -o output config

# or, just compute assembly metrics
$ mf metrics -o output config

# or, create the 'config' file in current directory
$ mf gen-conf config

```

Moreover, for single-cell paired-end data, the option '-sc' is required be set to 1, and the command will be:
```sh
$ mf all -sc 1 -o output config 

```

The help information can be shown:
```sh
misFinder: 0.5.0
Released : Feb 2, 2024

Usage: mf <command> [options] <CONF_FILE>

Description:
    CONF_FILE   Configuration file (required)

Commands:
    metrics     Compute the assembly metrics
    misass      Compute mis-assemblies
    all         Do all the above in turn
    gen-conf    Create configuration file

Options:
  1) metrics    -- compute the metrics:
    -m <INT>        The minimal query length (bp). [100]
    -pt <FLOAT>     The minimal identity percentage for matched queries and 
                    matched segments. [0.9]
    -t <INT>        The number of threads for the alignment between queries 
                    and subjects. Default is the number of CPU cores.
    -o <STR>
    -out <STR>      Output directory for the output files. [output]
    -h
    -help           Show help information.
  2) misass     -- compute mis-assemblies:
    -i <INT>        Minimal indel size (bp). [5]
    -t <INT>        The number of threads for reads alignment. Default is
                    the number of CPU cores.
    -sc <INT>       Single-cell paired-end data flag. [0]
                    0: standard genomic DNA prepared from culture;
                    1: single-cell data.
    -o <STR>
    -out <STR>      Output directory for the output files. [output]
    -h
    -help           Show help information.
  3) gen-conf   -- create configuration file in current directory:
    -h
    -help           Show help information.

Example:
  # detect mis-assemblies based on the specified information in 'config' file 
  $ mf all -o output config

  # just compute assembly metrics
  $ mf metrics -o output config

  # create the 'config' file in current directory
  $ mf gen-conf config
```


## Output Result Description

Then, misFinder will perform the BLASTN alignment, identify the assembly errors and correct assemblies due to structural variations, and then correct the assembly errors at their breakpoints. Finally, the final results will be output into five files, including four files for the identified mis-assembly information and one file for the corrected assemblies:
* result_errors:
    The identified assembly errors.
* result_sv:
    The correct assemblies due to structural variations.
* result_gap:
    The indel caused by gap regions filled by poly_N.
* result_warning:
    The warnings produced by misFinder, they may need further checking.
* query_cor.fa:
    The corrected assemblies after performing the correction at the breakpoints of assembly errors.


-------------

## Contact

If you have any problems, comments, or suggestions, please contact [xzhu@ytu.edu.cn](xzhu@ytu.edu.cn) without hesitation. Thank you very much!

---- Enjoy !!! -----

